finetune/siamese_comparison.py

```diff
@@ -0,0 +1,544 @@
+from finetune.imbalance import compute_class_weights, class_weight_tensor
+from finetune.errors import FinetuneError
+
+JL_BASE = os.path.join(os.path.dirname(__file__), "model", "Base_model.jl")
+
+_LOGGER = logging.getLogger(__name__)
+
+DROPOUT_ON = 1
+DROPOUT_OFF = 0
+SAVE_PREFIX = 'model'
+MIN_UPDATES = 15
+
+
+class SiameseComparison(BaseModel):
+    """
+    Compares two documents and produces a similarity score (between 0-1).
+
+    :param config: A :py:class:`finetune.config.Settings` object or None (for default config).
+    :param \**kwargs: key-value pairs of config items to override.
+    """
+
+    def _text_to_ids(self, pairs, Y=None, max_length=None):
+        """
+        Format comparison examples as a list of IDs
+
+        pairs: Array of text, shape [batch, 2]
+        """
+        return super()._text_to_ids(pairs[0], Y=Y, max_length=max_length), \
+               super()._text_to_ids(pairs[1], Y=Y, max_length=max_length)
+
+    def _construct_graph(self, n_updates_total, target_dim=None, train=True):
+        gpu_grads = []
+        self.summaries = []
+
+        # store whether or not graph was previously compiled with dropout
+        self.train = train
+        self._define_placeholders(target_dim=target_dim)
+
+        aggregator = defaultdict(list)
+        train_loss_tower = 0
+        gpus = self.config.visible_gpus
+        n_splits = max(len(gpus), 1)
+
+        # multi-GPU setup, using CPU as param server is most efficient unless system has direct GPU connections
+        # single GPU, no need to use a different GPU as a parameter server
+        params_device = 'cpu' if len(gpus) != 1 else gpus[0]
+
+        # decide on setting for language model loss coefficient
+        # if the language model loss does not contribute to overall loss,
+        # remove the language model computation from the graph
+        lm_loss_coef = self.config.lm_loss_coef
+        if target_dim is None:
+            lm_loss_coef = 1.0
+        compile_lm = (train and lm_loss_coef > 0) or self.require_lm
+
+        for i, (X1, M1, X2, M2, Y) in enumerate(soft_split(self.X1, self.M1, self.X2, self.M2, self.Y, n_splits=n_splits)):
+            do_reuse = True if i > 0 else tf.AUTO_REUSE
+
+            if gpus:
+                device = tf.device(assign_to_gpu(gpus[i], params_device=params_device))
+            else:
+                device = tf.device('cpu')
+
+            scope = tf.variable_scope(tf.get_variable_scope(), reuse=do_reuse)
+
+            with device, scope:
+                featurizer_state1 = featurizer(
+                    X1,
+                    config=self.config,
+                    encoder=self.encoder,
+                    dropout_placeholder=self.do_dropout,
+                    train=train,
+                    reuse=do_reuse
+                )
+
+                # TODO what does the reuse parameter do here?
+                # Setting `reuse=1` here assuming it means the siamese networks share weights...
```

Comment:
This is the reuse parameter that gets passed to `tf.variable_scope()` -- see https://www.tensorflow.org/guide/variables.